internal/backend/post.go

```diff
@@ -46,64 +48,74 @@ func (d *delphisBackend) CreatePost(ctx context.Context, discussionID string, pa
 		ImportedContentID: input.ImportedContentID,
 	}
 
-	// Begin tx
-	tx, err := d.db.BeginTx(ctx)
-	if err != nil {
-		logrus.WithError(err).Error("failed to begin tx")
-		return nil, err
-	}
-	// Put post contents
-	if err := d.db.PutPostContent(ctx, tx, postContent); err != nil {
-		logrus.WithError(err).Error("failed to PutPostContent")
-
-		// Rollback on errors
-		if txErr := d.db.RollbackTx(ctx, tx); txErr != nil {
-			logrus.WithError(txErr).Error("failed to rollback tx")
-			return nil, multierr.Append(err, txErr)
+	retryAttempts := 0
+	for true {
+		// Begin tx
+		tx, err := d.db.BeginTx(ctx)
+		if err != nil {
+			logrus.WithError(err).Error("failed to begin tx")
+			return nil, err
 		}
-		return nil, err
-	}
-
-	// Put post
-	postObj, err := d.db.PutPost(ctx, tx, post)
-	if err != nil {
-		logrus.WithError(err).Error("failed to PutPost")
-
-		// Rollback on errors
-		if txErr := d.db.RollbackTx(ctx, tx); txErr != nil {
-			logrus.WithError(txErr).Error("failed to rollback tx")
-			return nil, multierr.Append(err, txErr)
+		// Put post contents
+		if err := d.db.PutPostContent(ctx, tx, postContent); err != nil {
+			logrus.WithError(err).Error("failed to PutPostContent")
+
+			// Rollback on errors
+			if txErr := d.db.RollbackTx(ctx, tx); txErr != nil {
+				logrus.WithError(txErr).Error("failed to rollback tx")
+				return nil, multierr.Append(err, txErr)
+			}
+			return nil, err
 		}
-		return nil, err
-	}
 
-	// Put Activity
-	if err := d.db.PutActivity(ctx, tx, postObj); err != nil {
-		logrus.WithError(err).Error("failed to PutActivity")
+		// Put post
+		postObj, err := d.db.PutPost(ctx, tx, post)
+		if err != nil {
+			if retryAttempts < PutPostMaxRetry && strings.Contains(strings.ToLower(err.Error()), "duplicate key value violates unique constraint") {
```

Comment:
I think using the error code type could be a safer and more consistent route: https://github.com/lib/pq/blob/master/error.go#L178

Reply:
This is now be addressed. I tested it, it seems to be working.